tests/test_utils.py

```diff
@@ -824,3 +826,118 @@ def test_non_operations_in_list(self, arg):
+
+        qml.enable_tape()
+
+        dev = qml.device('default.qubit', wires=1)
+
+        @qml.qnode(dev)
+        def circuit(w):
+            qml.RX(w, wires=0)
+            qml.Hadamard(wires=0)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        with Spy() as counter:
+            # 2 executions
+            circuit(0.1)
+            circuit(0.5)
+
+            # 3 executions
+            g = qml.grad(circuit)
+            g(0.1)
+
+            # 1 execution
+            circuit.qtape.execute(dev)
+
+        assert counter.counts == 6
+
+        qml.disable_tape()
+
+    def test_multi_output(self):
+        """Tests a qnode with a 2-d output."""
+
+        qml.enable_tape()
+
+        dev = qml.device('default.qubit', wires=1)
+
+        @qml.qnode(dev)
+        def circuit(w):
+            qml.RX(w, wires=0)
+            qml.Hadamard(wires=0)
+            return qml.expval(qml.PauliZ(wires=0)), qml.expval(qml.PauliZ(wires=1))
+
+        with Spy() as counter:
+            # 2 executions
+            circuit(0.1)
+            circuit(0.5)
+
+            # 5 executions
+            g = qml.jacobian(circuit)
+            g(0.1)
+
+            # 1 execution
+            circuit.qtape.execute(dev)
+
+        assert counter.counts == 8
+
+        qml.disable_tape()
+
+    def test_exit_restores_function(self):
+        """Tests that the mock only applies in a context."""
+
+        qml.enable_tape()
+
+        dev = qml.device('default.qubit', wires=1)
+
+        @qml.qnode(dev)
+        def circuit(w):
+            qml.RX(w, wires=0)
+            qml.Hadamard(wires=0)
+            return qml.expval(qml.PauliZ(wires=0)), qml.expval(qml.PauliZ(wires=1))
+
+        with Spy():
+            circuit(0.1)
+            assert isinstance(qml._qubit_device.QubitDevice.execute, Mock)
+
+        assert not isinstance(qml._qubit_device.QubitDevice.execute, Mock)
+
+        qml.disable_tape()
+
+    def test_execute_not_run(self):
+        """Tests that in a mock context the execute function is never run."""
+
+        qml.enable_tape()
+
+        def exception_raiser(circuit, kwargs={}):
+            raise NotImplementedError
+
+        temp = qml._qubit_device.QubitDevice.execute
+        # replace the execute function with a function of same signature,
+        # but which always raises an exception
+        qml._qubit_device.QubitDevice.execute = exception_raiser
+
+        dev = qml.device('default.qubit', wires=1)
+
+        @qml.qnode(dev)
+        def circuit(w):
+            qml.RX(w, wires=0)
+            qml.Hadamard(wires=0)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        with Spy():
+            res = circuit(0.1)
+            # assert that we get the counter's dummy output
+            assert res == np.array(0.)
+
+        with pytest.raises(NotImplementedError):
+            circuit(0.1)
+
+        qml._qubit_device.QubitDevice.execute = temp
+        qml.disable_tape()
```

Comment:
Great tests!

Perhaps it might be useful to test other interfaces as well? I see no reason why something like

```python
with Spy() as spy:
    x = tf.Variable(0.1)
    with tf.GradientTape() as tape:
        res = circuit(0.1)
    tape.gradient(res, x)
```

shouldn't also work.